### ISSUE TITLE
Revert returning a single NaN for `np.unique` in 1.22.0?

### ISSUE THREAD

**Issue by rgommers** [numpy/numpy#20326]
[This mailing list thread](https://mail.python.org/archives/list/numpy-discussion@python.org/thread/DBJZNCNBXTNU2EWKDZMP7MDQLOC46GNV/?sort=thread) discusses reverting the change in `1.21.0` to make `unique` return a single `nan` in its output when the input array contains multiple `nan`s. This issue is to track that, needs a decision before `1.22.0rc1` I think.

**Comment by rgommers** [comment]
Also see gh-19655, which shows that `np.ma.unique` still returns multiple `nan`s in `1.21.0`.

**Comment by asmeurer** [comment]
If NumPy keeps its current behavior (single NaN), it would be useful to have an internal function that computes the old behavior, so that it can be used in the array_api submodule which requires the old, multiple NaN behavior. Otherwise it would have to workaround this in a way that involves building a new array from the result from `np.unique`. 

The inverse, if we do revert, could also be useful (or it could be added as a flag to `unique`), but wouldn't be required as far as array_api is concerned.

**Comment by asmeurer** [comment]
By the way, it would be useful to add a label in the issue tracker for numpy.array_api.

**Comment by rgommers** [comment]
> By the way, it would be useful to add a label in the issue tracker for numpy.array_api.

Done: https://github.com/numpy/numpy/pulls?q=is%3Aopen+is%3Apr+author%3Aasmeurer+label%3A%22component%3A+numpy.array_api%22

**Comment by charris** [comment]
@asmeurer I was thinking a  keyword would be useful, something line `equal_nans=True`. Note that if nans are ever made singletons, the old behavior will change anyway.

**Comment by rgommers** [comment]
> @asmeurer I was thinking a keyword would be useful, something line `equal_nans=True`.

That may be a good option if we have implementations for both behaviors and we know that there is demand for both from downstream libraries.

**Comment by rgommers** [comment]
R's `NA` is not `nan`, it's "missing value" - R has both and they're conceptually different.

**Comment by HaoZeke** [comment]
Sorry, my bad. Still the same behaviour though:

```R
> a=c(NaN,3,1,5,6,7,NaN,2,NaN,2)
> unique(a)
[1] NaN   3   1   5   6   7   2
```

**Comment by seberg** [comment]
In the meeting, we seemed to converge on: lets just add akwarg.  The not quite clear followup to that is what the default should be.

**Comment by rgommers** [comment]
Agreed with a kwarg, because there's a desire for both sets of behaviors. It's indeed unclear what the default should be. This was a silent change in behavior in `1.21.0`, which could have broken existing valid code (e.g. people chopping off the last values of some returned array, rather than post-processing with `np.isnan`). So if we had caught this sooner, it'd be pretty clear-cut - then the default value should have been backwards-compatible, i.e. `equal_nans=False`. 

Now that we're 4-5 months in though, it's more fuzzy. There may be people who updated their code, and changing it back would break them again. So I'd say perhaps we should leave it as it is in `1.21.0`.

Leaving this on the `1.22.0` milestone, we should add this option even if we only get to it post `.rc1` (or perhaps we should make sure to include it - @charris?).

**Comment by charris** [comment]
> we should add this option even if we only get to it post .rc1

Should not be a problem as long as there is a test for both behaviors.

**Comment by rgommers** [comment]
This will not make it into `1.22.0rc1`, so I suggest to stay with `equal_nans=True` as the default (matching 1.21.x behavior).

**Comment by asmeurer** [comment]
It should be noted that `numpy.array_api.unique` (and related unique functions) currently do not conform to the array API spec, because I didn't find it worth trying to work around NumPy's current behavior given that a flag is planned. So when we add the flag, we should be sure to also use it in the array API submodule so that it becomes correct (unique is not yet tested in the array-api-tests test suite, but it should be added soon).

**Comment by charris** [comment]
Pushing off to 1.23.

**Comment by charris** [comment]
Adding a keyword looks like a fairly simple enhancement that we should try to get into 1.23. At this point I think it preferable to keep the current behavior as the default.

**Comment by zhouhuaian** [comment]
👍 💯

**Comment by zhouhuaian** [comment]
😃
